### ISSUE TITLE
how to merge different Convolution2D layers when using the GRAPH model 

### ISSUE THREAD

**Issue by alyato** [fchollet/keras#2685]
I want to use the graph model to train my model and use the mnist dataset. Then i rewrite the example/mnist_cnn.py 
![gr](https://cloud.githubusercontent.com/assets/6815071/15147117/2ed8b2ee-167c-11e6-9e0e-f9ccefa38003.png)

Both pool2 layer and conv3 layer   as the input  in the flat1 layer. I set the merge_mode='concat'.
we can get the output 32x6x6 in conv3 layer and 32x3x3 in pool2 layer.
Then i set the value（32x9x9）in the dense1 layer.I think it's wrong and don't know how to set the values.
![qq](https://cloud.githubusercontent.com/assets/6815071/15147321/589e0768-167d-11e6-9e5b-12318aa9922b.png)

**Comment by kgrm** [comment]
The shapes of the layers you're trying to concatenate can only differ in one dimension - the dimension you're concatenating them along. For example, if you were trying to concatenate outputs of convolutional layers with different numbers of filters, they could differ along the axis 1, with shapes (N, P, x, y) and (N, Q, x, y), with the resulting concatenated output of shape (N, P+Q, x, y). As you can see, however, the tensors you're trying to concatenate differ in shape along two axes, which is invalid. Consider  cropping, padding or resampling one of them so their final two dimensions will be the same.

**Comment by alyato** [comment]
@kgrm i can't understand what you say completely.
I get the outputs of different convolutions in my code.
 conv1  layer is (32x24x24),conv2 layer is (32x20x20),and conv3 layer is (32x6x6),pool2 layer is (32x3x3). It is different between conv3 layer and pool2 layer though they have the some number of filters.(32x6x6 not equal 32x3x3)

> if you were trying to concatenate outputs of convolutional layers with different numbers of filters, they could differ along the axis 1, with shapes (N, P, x, y) and (N, Q, x, y), with the resulting concatenated output of shape (N, P+Q, x, y)

Above sentence, can i understand that N ==N ,x==x ,y==y ,but P==Q or P!=Q are ok.

> conv3 is 32x6x6 and pool2 is 32x3x3.(x=y=6 ) != (x=y=3).There will be wrong. So i must change the shape in conv3 or pool2,like (N,P,3,3) in conv3 or (N,Q,6,6).

**Comment by kgrm** [comment]
Yes, the point is, if you're trying to concatenate two tensors, their dimensions can only differ along a single axis, in which case the shape of the resulting tensor will be identical along other axes, but the sum of the sizes of the concatenated tensors along the concatenation axis.

**Comment by alyato** [comment]
@kgrm  thanks. And i read the torch code about the concat. It's called DepthConcat in torch.[https://github.com/torch/nn/blob/master/doc/containers.md#depthconcat](url).
It says that

> DepthConcat concatenates the output of one layer of "parallel" modules along the provided dimension dim: they take the same inputs, and their output is concatenated. For dimensions other than dim having different sizes, the smaller tensors are copied in the center of the output tensor, effectively padding the borders with zeros

I understand that my code can concatenate in torch,don't like the keras.It will auto add zero padding even though the different inputs.

> If i concatenate the (32x6x6) and (32x3x3) in torch ,the (32x3x3) will convert to the (32x6x6).Then (x=y=6)==(x=y=6)

But it can't do that auto and the N ==N ,x==x ,y==y  in keras. Is it right?

**Comment by shay86** [comment]
I built a project using (Merge, merge) layer once, and use share layer in another. The problem I can't find any documentation that explains how merge or share layer works like convolution or max-pooling layers.
I will be grateful if anyone can direct me or suggest some paper to me that helps me to understand how these layers work ???
